bitbots_body_behavior/src/bitbots_body_behavior/decisions/allowed_to_move.py

```diff
@@ -6,29 +6,38 @@
 class AllowedToMove(AbstractDecisionElement):
     def __init__(self, blackboard, dsd, parameters=None):
         super(AllowedToMove, self).__init__(blackboard, dsd, parameters)
+        self.ball_lost_time = rospy.Duration.from_sec(self.blackboard.config['ball_lost_time'])
+        self.kickoff_min_ball_movement = rospy.get_param("behavior/body/kickoff_min_ball_movement")
 
     def perform(self, reevaluate=False):
         """
         Determines whether the robot is penalized or was just unpenalized.
         :param reevaluate:
         :return:
         """
-        if self.blackboard.gamestate.get_is_penalized() or self.blackboard.gamestate.get_gamestate() == GameState.GAMESTATE_FINISHED:
+        if self.blackboard.gamestate.get_is_penalized() \
+                or self.blackboard.gamestate.get_gamestate() == GameState.GAMESTATE_FINISHED:
             return 'NO_MOVEMENT'
         elif self.blackboard.gamestate.get_gamestate() in (GameState.GAMESTATE_INITAL, GameState.GAMESTATE_SET):
             return 'ONLY_HEAD'
-        elif self.blackboard.gamestate.get_secondary_state() in (GameState.STATE_DIRECT_FREEKICK,
-                                                       GameState.STATE_INDIRECT_FREEKICK,
-                                                       GameState.STATE_PENALTYKICK,
-                                                       GameState.STATE_CORNER_KICK,
-                                                       GameState.STATE_GOAL_KICK,
-                                                       GameState.STATE_THROW_IN) and \
-                self.blackboard.gamestate.get_secondary_state_mode() in (GameState.MODE_PREPARATION, GameState.MODE_END):
+        elif self.blackboard.gamestate.get_secondary_state() != GameState.STATE_NORMAL \
+                and self.blackboard.gamestate.get_secondary_state_mode() in (GameState.MODE_PREPARATION,
+                                                                             GameState.MODE_END):
             return 'ONLY_HEAD'
         elif self.blackboard.gamestate.get_gamestate() == GameState.GAMESTATE_READY:
             return 'NORMAL'
         elif self.blackboard.gamestate.get_gamestate() == GameState.GAMESTATE_PLAYING:
-            if not self.blackboard.gamestate.has_kickoff() and self.blackboard.gamestate.get_secondary_seconds_remaining() != 0:
+            if not self.blackboard.gamestate.has_kickoff():
+                if self.blackboard.gamestate.get_secondary_seconds_remaining() == 0:
+                    # time is up for the other team
+                    return 'NORMAL'
+                ball_pos = self.blackboard.world_model.get_ball_position_xy()
+                # if we know where the ball is and that it moved, we can play too
+                if rospy.Time.now() - self.blackboard.world_model.ball_last_seen() < self.ball_lost_time and (
```

Comment:
Why do we check for a lost ball in the allowed-to-move decision? Normally this is done in the dsd file.

Reply:
I would add another decision for the ball moved.

Reply:
I would keep it for now like this. I agree that things could be improved, but there are a lot of things all over the body behavior which need to be refactored and I would rather like to do this after the competition.

Reply:
It now publishes additional debug data